platform/core/src/services/ToolBarService/ToolbarService.ts

```diff
@@ -130,160 +136,184 @@ export default class ToolbarService extends PubSubService {
-
-    // if not interaction type, assume the type can be used
-    if (!interactionType) {
-      interactionType = type;
+  ) {
+    // if interaction is a string, we can assume it is the itemId
+    // and get the props to get the other properties
+    if (typeof interaction === 'string') {
+      interaction = this.getButtonProps(interaction);
     }
 
-    switch (interactionType) {
-      case 'action': {
-        commandsManager.run(commands, options);
-        break;
-      }
-      case 'tool': {
-        try {
-          const alternateInteraction =
-            this.state.primaryToolId === itemId &&
-            this.defaultTool?.itemId !== itemId &&
-            this.getDefaultTool();
-          if (alternateInteraction) {
-            // Allow toggling the mode off
-            return this.recordInteraction(alternateInteraction, options);
-          }
-          commandsManager.run(commands, options);
-
-          // only set the primary tool if no error was thrown.
-          // if the itemId is not undefined use it; otherwise, set the first tool in
-          // the commands as the primary tool
-          this.state.primaryToolId = itemId || commands[0].commandOptions?.toolName;
-        } catch (error) {
-          console.warn(error);
-        }
-
-        break;
-      }
-      case 'toggle': {
-        const { commands } = interaction;
-        let commandExecuted;
-
-        // only toggle if a command was executed
-        this.state.toggles[itemId] =
-          this.state.toggles[itemId] === undefined ? true : !this.state.toggles[itemId];
+    const itemId = interaction.itemId ?? interaction.id;
+    interaction.itemId = itemId;
 
-        if (!commands) {
-          break;
-        }
+    const commands = Array.isArray(interaction.commands)
+      ? interaction.commands
+      : [interaction.commands];
 
-        commands.forEach(({ commandName, commandOptions, context }) => {
-          if (!commandOptions) {
-            commandOptions = {};
-          }
-
-          if (commandName) {
-            commandOptions.toggledState = this.state.toggles[itemId];
-
-            try {
-              commandsManager.runCommand(commandName, commandOptions, context);
-              commandExecuted = true;
-            } catch (error) {
-              console.warn(error);
-            }
-          }
-        });
-
-        if (!commandExecuted) {
-          // If no command was executed, we need to toggle the state back
-          this.state.toggles[itemId] = !this.state.toggles[itemId];
-        }
+    if (!commands?.length) {
+      this.refreshToolbarState({
+        ...options?.refreshProps,
+        itemId,
+        interaction,
+      });
+    }
 
-        break;
+    // attach interaction to each command if it is an object {commandName, commandOptions}
+    commands.forEach(command => {
+      if (typeof command === 'object') {
+        command.commandOptions = { ...command.commandOptions, ...options, ...interaction };
```

Comment:
You should construct the options outside the commands loop, not containing the commandOptions, and let commandsManager.run(command,options) combine the options as needed.  Then you can just:
this._commandsManager.run(commands, {...options, ...interaction});